src/components/webview/MainWebviewProvider.ts

```diff
@@ -115,160 +122,117 @@ export class IntuitaProvider implements WebviewViewProvider {
-			];
+	private __getTreeByDirectory = (element: Element): TreeNode | undefined => {
+		const mappedNode: TreeNode = {
+			id: element.hash,
+			iconName: getElementIconBaseName(element.kind),
+		};
 
-			if (this.__jobManager.isJobAccepted(element.jobHash)) {
-				mappedNode.kind = 'acceptedJobElement';
-
-				mappedNode.actions = [
-					{
-						title: '✗ Dismiss',
-						command: 'intuita.rejectCase',
-						arguments: [element.hash],
-					},
-					{
-						title: 'Issue',
-						command: 'intuita.createIssue',
-						arguments: [element.job.hash],
-					},
-					{
-						title: 'PR',
-						command: 'intuita.createPR',
-						arguments: [element.job.hash],
-					},
-				];
-			}
+		if (element.kind === ElementKind.JOB) {
+			this.__buildJobTree(element, mappedNode);
+			return {
+				...mappedNode,
+				label:
+					jobKindCopyMap[element.job.kind] +
+					` from ${element.job.codemodName}`,
+			};
 		}
 
-		if (element.kind === 'CASE') {
-			mappedNode.kind = 'caseElement';
-			mappedNode.actions = [
-				{
-					title: '✓ Apply',
-					command: 'intuita.acceptCase',
-					arguments: [element.hash],
-				},
-				{
-					title: '✗ Dismiss',
-					command: 'intuita.rejectCase',
-					arguments: [element.hash],
-				},
-			];
-
-			const caseJobHashes = this.__caseManager.getJobHashes([
-				String(element.hash) as CaseHash,
-			]);
-			const caseAccepted = Array.from(caseJobHashes).every((jobHash) =>
-				this.__jobManager.isJobAccepted(jobHash),
-			);
+		if (element.kind === ElementKind.ROOT) {
+			element.children.forEach(this.__getTreeByDirectory);
+			return {
+				...mappedNode,
+				label: 'Recipe',
+				children: this.__folderMap.has(ROOT_FOLDER_KEY)
+					? [this.__folderMap.get(ROOT_FOLDER_KEY) as TreeNode]
+					: [],
+			};
+		}
 
-			if (caseAccepted) {
-				mappedNode.kind = 'acceptedCaseElement';
-
-				mappedNode.actions = [
-					{
-						title: '✗ Dismiss',
-						command: 'intuita.rejectCase',
-						arguments: [element.hash],
-					},
-					{
-						title: 'Issue',
-						command: 'intuita.createIssue',
-						arguments: [element.hash],
-					},
-					{
-						title: 'PR',
-						command: 'intuita.createPR',
-						arguments: [element.hash],
-					},
-				];
-			}
+		if (element.kind === ElementKind.CASE) {
+			element.children.forEach(this.__getTreeByDirectory);
 		}
 
-		mappedNode.children =
-			'children' in element ? element.children.map(this.__getTree) : [];
+		if (element.kind === ElementKind.FILE) {
+			// e.g., extract the path from '/packages/app/src/index.tsx (1)'
+			const filePath = element.label.split(' ')[0];
+			if (!filePath) return;
+
+			// e.g., ['packages', 'app', 'src', 'index.tsx']
+			let directories = filePath.split('/').filter((item) => item !== '');
+			// e.g., ['root', 'packages', 'app', 'src', 'index.tsx']
+			directories = ['root', ...directories];
+			// e.g., 'index.tsx'
+			const fileName = directories[directories.length - 1];
+			let path = '';
+			let node: TreeNode = { id: '' };
```

Comment:
I'd create and return `node` when needed

Reply:
Can you elaborate?

Reply:
Ideally, when you have a modifiable object, you should keep the modifications scoped, so everybody can see where it's changed with a glimpse of an eye. That's why I prefer to use immutable objects and just create the objects when they are needed in-place.